### PR TITLE
Implement `Smt` struct (replacement to `TieredSmt`)

### PR DESCRIPTION
**Pull request by plafer**
Builds on #245

### CODE REVIEW

File: src/merkle/smt/full/tests.rs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        assert_eq!(smt.root(), tree_root);
+    }
+}
+
+/// This test checks that inserting twice at the same key functions as expected. The test covers
+/// only the case where the leaf type is `SmtLeaf::Multiple`
+#[test]
+fn test_smt_insert_at_same_key_2() {
+    let key_already_present: RpoDigest = {
+        let raw = 0b_01101001_01101100_00011111_11111111_10010110_10010011_11100000_00000000_u64;
```
Comment (bobbinth) [inline src/merkle/smt/full/tests.rs:54]: nit: I think this is the same value as the one on line 75 below (it took me some time to understand this). I'm not sure there is a big benefit to using binary representation here and it actually makes this a bit more difficult to follow. I'd probably use just decimal values for this.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left some minor comments inline. After these are addressed, we are good to merge.

File: src/merkle/smt/full/mod.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct Smt {
+    root: RpoDigest,
+    leaves: BTreeMap<u64, SmtLeaf>,
+    inner_nodes: BTreeMap<NodeIndex, InnerNode>,
+}
+
+impl Smt {
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs:37]: done

File: src/merkle/smt/full/mod.rs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    pub fn open(&self, key: &RpoDigest) -> (MerklePath, SmtLeaf) {
+        <Self as SparseMerkleTree<SMT_DEPTH>>::open(self, key)
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts a value at the specified key, returning the previous value associated with that key.
+    /// Recall that by definition, any key that hasn't been updated is associated with
+    /// [`EMPTY_WORD`].
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: done

File: src/merkle/smt/full/mod.rs [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+impl Smt {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a new [NewSmt].
+    ///
+    /// All leaves in the returned tree are set to [ZERO; 4].
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: done

File: src/merkle/smt/full/mod.rs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+    /// Returns a new [SimpleSmt] instantiated with leaves set as specified by the provided entries.
+    ///
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number of entries exceeds 2^63 entries.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: done

File: src/merkle/smt/full/mod.rs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    /// Remove key-value pair into the leaf stored at key; return the previous value associated with
+    /// `key`, if any. We also return an `is_empty` flag which indicates whether the leaf became
+    /// empty, and must be removed from the data structure it is contained in.
+    fn remove(&mut self, key: RpoDigest) -> (Option<Word>, bool) {
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: fixed docstring

File: src/merkle/smt/full/mod.rs [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+            }
+        }
+    }
+
+    // HELPERS
+    // ---------------------------------------------------------------------------------------------
+
+    /// Insert key-value pair into the leaf; return the previous value associated with `key`, if
+    /// any.
+    fn insert(&mut self, key: RpoDigest, value: Word) -> Option<Word> {
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: done

File: src/merkle/smt/full/mod.rs [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    /// Converts a leaf to a list of field elements
+    pub fn into_elements(self) -> Vec<Felt> {
+        match self {
+            SmtLeaf::Single(kv_pair) => kv_to_elements(kv_pair).collect(),
+            SmtLeaf::Multiple(kv_pairs) => kv_pairs.into_iter().flat_map(kv_to_elements).collect(),
+        }
+    }
+
+    /// Compute the hash of the leaf
+    pub fn hash(&self) -> RpoDigest {
```
[the remark this answers is not in the document]
Comment (plafer) [reply src/merkle/smt/full/mod.rs]: done

File: src/merkle/smt/full/tests.rs (hunk 1 above)
Comment (plafer) [reply src/merkle/smt/full/tests.rs]: right - cleaned it up

**Comment by plafer** [comment]
@bobbinth addressed PR comments, and fixed `SmtLeaf` as discussed in [a880c88](https://github.com/0xPolygonMiden/crypto/pull/254/commits/a880c880ab4f9f22507de24536b99f0700f61ebb)

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**